fix(utils): respect patterns within paths when sorting routes

marcelobotega opened the pull request:
<!---
☝️ PR title should follow conventional commits (https://conventionalcommits.org)

Please carefully read the contribution docs before creating a pull request
 👉 https://nuxt.com/docs/community/contribution
-->

### 🔗 Linked issue

Fixes #10630 

### ❓ Type of change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->

- [ ] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

Update sortRoutes to handle routes with letters and * (test*)
Resolves #10630 
<!-- Describe your changes in detail -->
<!-- Why is this change required? What problem does it solve? -->
<!-- If it resolves an open issue, please link to the issue here. For example "Resolves #1337" -->

### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

marcelobotega committed:
fix: update sortRoutes to sort * with letters

marcelobotega committed:
fix: update sortRoutes regex to sort * with letters

marcelobotega committed:
test: add sort route fixture

marcelobotega committed:
fix: add a new level to taken the / into account

marcelobotega committed:
test: add fixture for route sort with slash

marcelobotega committed:
test: update snapshot

marcelobotega committed:
Merge branch '2.x' into bugfix/sorte-route-fix

# Conflicts:
#	packages/utils/test/__snapshots__/route.test.js.snap
#	yarn.lock

marcelobotega committed:
test: update snapshot

marcelobotega committed:
chore: update yarn.lock

marcelobotega committed:
chore: rename test file

marcelobotega committed:
test:  remove file with * that is not allowed

marcelobotega committed:
refactor: remove regex change and the new level on the ternary if

marcelobotega committed:
test: add posix to sortRoute unit test

danielroe commented on package.json:
```diff
@@ -38,6 +38,7 @@
     "@ls-lint/ls-lint": "^1.11.2",
     "@nuxtjs/eslint-config": "^12.0.0",
     "@nuxtjs/eslint-config-typescript": "^12.0.0",
+    "@nuxtjs/i18n": "^7.3.1",
```
do you think we could adequately test this change without adding the module as a dependency?

marcelobotega replied:
Without the module we can't generate the fixtures to test the "real" problem 😞 
However, I could try something, but not sure if it is the best solution. 
We can utilize both the routes.json files from the fixtures, unsort them, and apply them to new unit tests to evaluate the sortRoute method directly.
Would this be better @danielroe ?

danielroe replied:
Yes, I think that would be fine 👍

marcelobotega committed:
refactor: update dynamic route regex

Make the / optional and match : or * anywhere in the string. This ensures that special routes will be below normal routes. Also ensures that route / will be sorted before the special routes.

marcelobotega committed:
refactor: remove fixtures and package nuxtjs/i18n in favor of unit tests

marcelobotega committed:
test: add sortRoute unit tests

marcelobotega committed:
test: update windows snapshot

marcelobotega committed:
Merge branch '2.x' into bugfix/sorte-route-fix

marcelobotega committed:
Merge branch '2.x' into bugfix/sorte-route-fix

marcelobotega committed:
Merge branch '2.x' into bugfix/sorte-route-fix

marcelobotega committed:
Merge branch '2.x' into bugfix/sorte-route-fix

marcelobotega commented:
@danielroe can we run the actions on this PR ?

marcelobotega committed:
Merge branch '2.x' into bugfix/sorte-route-fix

marcelobotega commented:
@danielroe we got an error on the windows audit, how can we fix that ? It's a timeout error.

rchl commented on packages/utils/test/route.test.js:
```diff
@@ -237,4 +237,109 @@ describe('util: route', () => {
+          { path: '/de/tech' },
+          { path: '/' },
+          { path: '/de/:post' },
+          { path: '/:post' },
+          { path: '/de*' },
```
Looks like `/de*` would be inaccessible this way? Should be before `/:post`?

marcelobotega replied:
Nice catch, yeah, the /de* won't work on this scenario. Not sure how relevant is this to the use case, but I'll try to fix it and have the same output as the `Should sortRoutes with extendRoutes using /*` test case.

marcelobotega committed:
fix(utils): update route order check to take only * into account and change the priority of param route

rchl reviewed (dismissed):
I've tested this on a couple of projects that (admittedly) only have a single `*` route but I haven't noticed any issues.

One thing that happens consistently is that the root route (`/`) changes position to after all other dynamic routes (ones that use `:`). So in theory that could result in slightly slower resolving of that particular route but not sure if it's something to be concerned about.

So thumbs up from me.

marcelobotega committed:
fix: update regex to be more efficient

Co-authored-by: Rafał Chłodnicki <rchl2k@gmail.com>

marcelobotega committed:
Merge branch '2.x' into bugfix/sorte-route-fix

danielroe committed:
Merge remote-tracking branch 'origin/2.x' into bugfix/sorte-route-fix

danielroe committed:
Merge remote-tracking branch 'origin/2.x' into bugfix/sorte-route-fix

danielroe commented on packages/utils/test/route.test.js:
```diff
@@ -238,4 +238,109 @@ describe('util: route', () => {
+      'pages/_param.vue',
+      'pages/de/index.vue',
+      'pages/index.vue',
+      'pages/*.vue',
```
shouldn't this (and other occurrences) be `_.vue` instead?

marcelobotega replied:
If I'm not mistaken, nuxt uses the `_.vue` as a catch-all, right ?
I changed here and when using `_` it goes back to the initial problem.
I found out the [getRoutePathExtension](https://github.com/nuxt/nuxt/blob/696dbc68447c95527c0d5dc2c324d5f93e4dc0cf/packages/utils/src/route.js#LL243C7-L243C28) method, and I am adding a snippet there to check the `_` in the end of the key. This should fix the problem for this use case.

rchl replied:
Actually I think you should not be doing anything with `_` that is neither the only character nor followed by some text.

You can test it yourself by creating `pages/test_.vue` file. Nuxt doesn't convert the `_` to `*`. So it's supposed to be handled as literal `_`.

marcelobotega replied:
Hey @rchl, 
I tested and if I add the route `de_.vue` we go back to the initial problem. The `de_` route goes to the top of the list and catches all `de` routes. 
I changed the [getRoutePathExtension](https://github.com/nuxt/nuxt/blob/d24275a1dc85b5be29a49cc23b66a1eab2f94552/packages/utils/src/route.js#L243) to handle the `_` in the end of the string (now it converts to *) and with that change everything seems to be right.

rchl replied:
Just test manually by creating a `de_.vue` file. Nuxt generates this route:
```
{
    path: "/de_",
    component: _0b861482,
    name: "de_"
  }
```

where `_` is treated literally and not converted to `*`. So it's fine if it's at the beginning and it shouldn't cause any issues because it's a literal match.

marcelobotega replied:
Ok, I understood now. 
My only question now is: Should the `_` in the end of the value be treated as a literal or not ?
For me, it makes sense to be treated as a special character, but I'm not sure if Nuxt should handle it like that, either way seems fine to me.
@danielroe can you give your insights on it ?

rchl replied:
[File-based routing in Nuxt](https://v2.nuxt.com/docs/features/file-system-routing) only supports the case when key is equal `_` or is followed by some characters. But modules can add routes with `*` anywhere in fact and the user can also do it using the [extendRoutes](https://v2.nuxt.com/docs/features/file-system-routing#extendroutes) option. So sorting of those should IMO still be supported but the changes should not affect the capabilities of the file-based routing (and I think your change in `getRoutePathExtension` would affect those).

marcelobotega replied:
Yes, it affects those. I'll roll back the commit with changes on the `getRoutePathExtension`

marcelobotega committed:
fix(utils): update getRoutePathExtension to check for _ in the end of the key

marcelobotega committed:
Merge branch '2.x' into bugfix/sorte-route-fix

marcelobotega committed:
revert(utils): update getRoutePathExtension to check for _ in the end of the key

marcelobotega committed:
Merge branch '2.x' into bugfix/sorte-route-fix
